src/main/java/com/yelp/nrtsearch/server/luceneserver/innerhit/InnerHitFetchTask.java

```diff
@@ -0,0 +1,111 @@
+import com.yelp.nrtsearch.server.grpc.SearchResponse.Diagnostics;
+import com.yelp.nrtsearch.server.grpc.SearchResponse.Diagnostics.Builder;
+import com.yelp.nrtsearch.server.grpc.SearchResponse.Hit;
+import com.yelp.nrtsearch.server.grpc.TotalHits;
+import com.yelp.nrtsearch.server.luceneserver.SearchHandler;
+import com.yelp.nrtsearch.server.luceneserver.search.FetchTasks.FetchTask;
+import com.yelp.nrtsearch.server.luceneserver.search.SearchContext;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.atomic.DoubleAdder;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.search.join.ParentChildrenBlockJoinQuery;
+
+public class InnerHitFetchTask implements FetchTask {
+  private static final double TEN_TO_THE_POWER_SIX = Math.pow(10, 6);
+
+  public InnerHitContext getInnerHitContext() {
+    return innerHitContext;
+  }
+
+  private final InnerHitContext innerHitContext;
+
+  private final DoubleAdder getFieldsTimeMs = new DoubleAdder();
+  private final DoubleAdder firstPassSearchTimeMs = new DoubleAdder();
+
+  public InnerHitFetchTask(InnerHitContext innerHitContext) {
+    this.innerHitContext = innerHitContext;
+  }
+
+  public void processHit(
+      SearchContext searchContext, LeafReaderContext hitLeaf, SearchResponse.Hit.Builder hit)
+      throws IOException {
+    long startTime = System.nanoTime();
+
+    ParentChildrenBlockJoinQuery parentChildrenBlockJoinQuery =
+        new ParentChildrenBlockJoinQuery(
+            innerHitContext.getParentFilter(), innerHitContext.getQuery(), hit.getLuceneDocId());
+
+    IndexSearcher searcher = innerHitContext.getSearcherAndTaxonomy().searcher;
+    TopDocs topDocs =
+        searcher.search(parentChildrenBlockJoinQuery, innerHitContext.getCollector().getManager());
+    if (innerHitContext.getStartHit() > 0) {
+      topDocs =
+          SearchHandler.getHitsFromOffset(
+              topDocs, innerHitContext.getStartHit(), innerHitContext.getTopHits());
+    }
+    firstPassSearchTimeMs.add(((System.nanoTime() - startTime) / TEN_TO_THE_POWER_SIX));
+
+    startTime = System.nanoTime();
+    HitsResult.Builder innerHitResultBuilder = HitsResult.newBuilder();
+    TotalHits totalInnerHits =
+        TotalHits.newBuilder()
+            .setRelation(TotalHits.Relation.valueOf(topDocs.totalHits.relation.name()))
+            .setValue(topDocs.totalHits.value)
+            .build();
+    innerHitResultBuilder.setTotalHits(totalInnerHits);
+    for (int hitIndex = 0; hitIndex < topDocs.scoreDocs.length; hitIndex++) {
+      var hitResponse = innerHitResultBuilder.addHitsBuilder();
+      ScoreDoc innerHit = topDocs.scoreDocs[hitIndex];
+      hitResponse.setLuceneDocId(innerHit.doc);
+      hitResponse.setScore(innerHit.score);
```

Comment:
This scoring information will be different based on if relevance or sorting is used, see https://github.com/Yelp/nrtsearch/blob/master/src/main/java/com/yelp/nrtsearch/server/luceneserver/search/collectors/additional/TopHitsCollectorManager.java#L182